### PR TITLE
python312Packages.pony: fix python3.12 compat, fix flask-mailman, fix flask-security-too

### PR DESCRIPTION
**Pull request by gador**
## Description of changes

- Pull upstream PR to fix `python3.12` compatibility.
- Update `flask-mailman` to fix compatibility with `python3.12` (found during nixpkgs-review)
- Fix `flask-security-too` by relaxing a dependency (failure found during nixpkgs-review)

Since these new build failures are only apparent when pony is fixed and are downstream, they are included in this PR

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

closes https://github.com/NixOS/nixpkgs/pull/326096

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by gador** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).

Result of `nixpkgs-review pr 325181` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>

  <summary>3 packages failed to build and are new build failures:</summary>
  <ul>
<li>agda-pkg: <a href="https://termbin.com/zvul">plain log</a></li>
<li>agda-pkg.dist: <a href="https://termbin.com/vztb">plain log</a></li>
<li>tribler: log was empty</li>
  </ul>
</details>
<details>
<summary>
22 packages built:
</summary>
<ul>
<li>mutmut</li>
<li>mutmut.dist</li>
<li>pgadmin4</li>
<li>pgadmin4-desktopmode</li>
<li>pgadmin4-desktopmode.dist</li>
<li>pgadmin4.dist</li>
<li>python311Packages.flask-mailman</li>
<li>python311Packages.flask-mailman.dist</li>
<li>python311Packages.flask-security-too</li>
<li>python311Packages.flask-security-too.dist</li>
<li>python311Packages.pony</li>
<li>python311Packages.pony.dist</li>
<li>python311Packages.ponywhoosh</li>
<li>python311Packages.ponywhoosh.dist</li>
<li>python312Packages.flask-mailman</li>
<li>python312Packages.flask-mailman.dist</li>
<li>python312Packages.flask-security-too</li>
<li>python312Packages.flask-security-too.dist</li>
<li>python312Packages.pony</li>
<li>python312Packages.pony.dist</li>
<li>python312Packages.ponywhoosh</li>
<li>python312Packages.ponywhoosh.dist</li>
</ul>
</details>

The following issues got detected with the above build packages.
Please fix at least the ones listed with your changed packages:

<details>
python311Packages.ponywhoosh.dist:
<details>

got build log for '/nix/store/h91ma0rjsc3kbxkdw6cnvvrn5nszn1dq-python3.11-ponywhoosh-1.7.8-dist' from 'daemon'
When evaluating attribute ‘python311Packages.ponywhoosh.dist’:
warning: no-python-tests
Test runner could not discover any test cases: ‘Ran 0 tests in 0.000s’
Near pkgs/development/python-modules/ponywhoosh/default.nix:26:

```
   |
26 |     description = "Make your database over PonyORM searchable";
   | ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/no-python-tests.md
</details>
python311Packages.ponywhoosh:
<details>

got build log for '/nix/store/s7x8vzwnidih2pyddiv3nsyc9waflq87-python3.11-ponywhoosh-1.7.8' from 'daemon'
When evaluating attribute ‘python311Packages.ponywhoosh’:
warning: no-python-tests
Test runner could not discover any test cases: ‘Ran 0 tests in 0.000s’
Near pkgs/development/python-modules/ponywhoosh/default.nix:26:

```
   |
26 |     description = "Make your database over PonyORM searchable";
   | ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/no-python-tests.md
</details>
python312Packages.ponywhoosh.dist:
<details>

got build log for '/nix/store/p0grs3dhqsyjvqkgn5vljqwdl3kndrj9-python3.12-ponywhoosh-1.7.8-dist' from 'daemon'
When evaluating attribute ‘python312Packages.ponywhoosh.dist’:
warning: no-python-tests
Test runner could not discover any test cases: ‘Ran 0 tests in 0.000s’
Near pkgs/development/python-modules/ponywhoosh/default.nix:26:

```
   |
26 |     description = "Make your database over PonyORM searchable";
   | ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/no-python-tests.md
</details>
python312Packages.ponywhoosh:
<details>

got build log for '/nix/store/jkirpmb8vkv9g8di671pbmcivig0fbj1-python3.12-ponywhoosh-1.7.8' from 'daemon'
When evaluating attribute ‘python312Packages.ponywhoosh’:
warning: no-python-tests
Test runner could not discover any test cases: ‘Ran 0 tests in 0.000s’
Near pkgs/development/python-modules/ponywhoosh/default.nix:26:

```
   |
26 |     description = "Make your database over PonyORM searchable";
   | ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/no-python-tests.md
</details>
</details>

**Comment by gador** [comment]
> 3 packages failed to build and are new build failures:
> 
>     * agda-pkg: [plain log](https://termbin.com/zvul)
> 
>     * agda-pkg.dist: [plain log](https://termbin.com/vztb)
> 

Python3.12 related (ModuleNotFoundError: No module named 'distutils')

>     * tribler: log was empty
> 

Probably also Python3.12 related

**Review by bendlas — APPROVED** [review]
This fixes pgadmin for me

File: pkgs/development/python-modules/pony/default.nix [hunk 1]
```diff
@@ -4,23 +4,31 @@
...
   src = fetchFromGitHub {
     owner = "ponyorm";
     repo = "pony";
     rev = "refs/tags/v${version}";
     hash = "sha256-wBqw+YHKlxYplgsYL1pbkusHyPfCaVPcH/Yku6WDYbE=";
   };
+  patches = [
+    # https://github.com/ponyorm/pony/pull/713
+    (fetchpatch {
+      name = "py12-compat.patch";
```
Comment (dotlambda) [inline pkgs/development/python-modules/pony/default.nix:27]:
```suggestion
      name = "py312-compat.patch";
```

File: pkgs/development/python-modules/pony/default.nix [hunk 2]
```diff
@@ -4,23 +4,31 @@
...
+      url = "https://github.com/ponyorm/pony/pull/713/commits/5a37f6d59b6433d17d6d56b54f3726190e98c98f.patch";
```
Comment (dotlambda) [inline pkgs/development/python-modules/pony/default.nix:28]:
```suggestion
      url = "https://github.com/ponyorm/pony/commit/5a37f6d59b6433d17d6d56b54f3726190e98c98f.patch";
```

File: pkgs/development/python-modules/pony/default.nix [hunk 3]
```diff
@@ -4,23 +4,31 @@
...
-  disabled = pythonOlder "3.8" || pythonAtLeast "3.12";
+  disabled = pythonOlder "3.8";
 
   src = fetchFromGitHub {
     owner = "ponyorm";
     repo = "pony";
     rev = "refs/tags/v${version}";
     hash = "sha256-wBqw+YHKlxYplgsYL1pbkusHyPfCaVPcH/Yku6WDYbE=";
   };
+  patches = [
```
Comment (dotlambda) [inline pkgs/development/python-modules/pony/default.nix:24]:
```suggestion

  patches = [
```

**Comment by superherointj** [comment]
Git commit body is missing release notes for:

> python312.flask-mailman: 1.1.0 -> 1.1.1

**Comment by superherointj** [comment]
Result of `nixpkgs-review pr 325181` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages failed to build:</summary>
  <ul>
    <li>agda-pkg</li>
    <li>agda-pkg.dist</li>
    <li>tribler</li>
  </ul>
</details>
<details>
  <summary>22 packages built:</summary>
  <ul>
    <li>mutmut</li>
    <li>mutmut.dist</li>
    <li>pgadmin4</li>
    <li>pgadmin4-desktopmode</li>
    <li>pgadmin4-desktopmode.dist</li>
    <li>pgadmin4.dist</li>
    <li>python311Packages.flask-mailman</li>
    <li>python311Packages.flask-mailman.dist</li>
    <li>python311Packages.flask-security-too</li>
    <li>python311Packages.flask-security-too.dist</li>
    <li>python311Packages.pony</li>
    <li>python311Packages.pony.dist</li>
    <li>python311Packages.ponywhoosh</li>
    <li>python311Packages.ponywhoosh.dist</li>
    <li>python312Packages.flask-mailman</li>
    <li>python312Packages.flask-mailman.dist</li>
    <li>python312Packages.flask-security-too</li>
    <li>python312Packages.flask-security-too.dist</li>
    <li>python312Packages.pony</li>
    <li>python312Packages.pony.dist</li>
    <li>python312Packages.ponywhoosh</li>
    <li>python312Packages.ponywhoosh.dist</li>
  </ul>
</details>

**Comment by superherointj** [comment]
Result of `nixpkgs-review pr 325181` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>14 packages failed to build:</summary>
  <ul>
    <li>agda-pkg</li>
    <li>agda-pkg.dist</li>
    <li>pgadmin4</li>
    <li>pgadmin4-desktopmode</li>
    <li>pgadmin4-desktopmode.dist</li>
    <li>pgadmin4.dist</li>
    <li>python311Packages.flask-mailman</li>
    <li>python311Packages.flask-mailman.dist</li>
    <li>python311Packages.flask-security-too</li>
    <li>python311Packages.flask-security-too.dist</li>
    <li>python312Packages.flask-mailman</li>
    <li>python312Packages.flask-mailman.dist</li>
    <li>python312Packages.flask-security-too</li>
    <li>python312Packages.flask-security-too.dist</li>
  </ul>
</details>
<details>
  <summary>10 packages built:</summary>
  <ul>
    <li>mutmut</li>
    <li>mutmut.dist</li>
    <li>python311Packages.pony</li>
    <li>python311Packages.pony.dist</li>
    <li>python311Packages.ponywhoosh</li>
    <li>python311Packages.ponywhoosh.dist</li>
    <li>python312Packages.pony</li>
    <li>python312Packages.pony.dist</li>
    <li>python312Packages.ponywhoosh</li>
    <li>python312Packages.ponywhoosh.dist</li>
  </ul>
</details>

**Comment by superherointj** [comment]
Darwin build is failing:
- python311Packages.flask-mailman: https://termbin.com/fbax

**Comment by gador** [comment]
Result of `nixpkgs-review pr 325181` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>agda-pkg</li>
    <li>agda-pkg.dist</li>
  </ul>
</details>
<details>
  <summary>22 packages built:</summary>
  <ul>
    <li>mutmut</li>
    <li>mutmut.dist</li>
    <li>pgadmin4</li>
    <li>pgadmin4-desktopmode</li>
    <li>pgadmin4-desktopmode.dist</li>
    <li>pgadmin4.dist</li>
    <li>python311Packages.flask-mailman</li>
    <li>python311Packages.flask-mailman.dist</li>
    <li>python311Packages.flask-security-too</li>
    <li>python311Packages.flask-security-too.dist</li>
    <li>python311Packages.pony</li>
    <li>python311Packages.pony.dist</li>
    <li>python311Packages.ponywhoosh</li>
    <li>python311Packages.ponywhoosh.dist</li>
    <li>python312Packages.flask-mailman</li>
    <li>python312Packages.flask-mailman.dist</li>
    <li>python312Packages.flask-security-too</li>
    <li>python312Packages.flask-security-too.dist</li>
    <li>python312Packages.pony</li>
    <li>python312Packages.pony.dist</li>
    <li>python312Packages.ponywhoosh</li>
    <li>python312Packages.ponywhoosh.dist</li>
  </ul>
</details>

**Comment by gador** [comment]
@superherointj thanks for the build and review! I was unable to reproduce your build failure for some reason?

**Comment by superherointj** [comment]
`agda-pkg` is just:

> ModuleNotFoundError: No module named 'distutils'

**Comment by superherointj** [comment]
Result of `nixpkgs-review pr 325181` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>14 packages failed to build:</summary>
  <ul>
    <li>agda-pkg</li>
    <li>agda-pkg.dist</li>
    <li>pgadmin4</li>
    <li>pgadmin4-desktopmode</li>
    <li>pgadmin4-desktopmode.dist</li>
    <li>pgadmin4.dist</li>
    <li>python311Packages.flask-mailman</li>
    <li>python311Packages.flask-mailman.dist</li>
    <li>python311Packages.flask-security-too</li>
    <li>python311Packages.flask-security-too.dist</li>
    <li>python312Packages.flask-mailman</li>
    <li>python312Packages.flask-mailman.dist</li>
    <li>python312Packages.flask-security-too</li>
    <li>python312Packages.flask-security-too.dist</li>
  </ul>
</details>
<details>
  <summary>10 packages built:</summary>
  <ul>
    <li>mutmut</li>
    <li>mutmut.dist</li>
    <li>python311Packages.pony</li>
    <li>python311Packages.pony.dist</li>
    <li>python311Packages.ponywhoosh</li>
    <li>python311Packages.ponywhoosh.dist</li>
    <li>python312Packages.pony</li>
    <li>python312Packages.pony.dist</li>
    <li>python312Packages.ponywhoosh</li>
    <li>python312Packages.ponywhoosh.dist</li>
  </ul>
</details>

**Comment by superherointj** [comment]
> @superherointj thanks for the build and review! I was unable to reproduce your build failure for some reason?

Might host might be unreliable. It's the shared community darwin builder. And I'm a bit too locked there to try to sort this out.

I will trust your host then.

Can you just fix `agda-pkg`? Seems straightforward.
I'd prefer zeroing the builds failures.

**Comment by superherointj** [comment]
Result of `nixpkgs-review pr 325181` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages failed to build:</summary>
  <ul>
    <li>agda-pkg</li>
    <li>agda-pkg.dist</li>
    <li>tribler</li>
  </ul>
</details>
<details>
  <summary>22 packages built:</summary>
  <ul>
    <li>mutmut</li>
    <li>mutmut.dist</li>
    <li>pgadmin4</li>
    <li>pgadmin4-desktopmode</li>
    <li>pgadmin4-desktopmode.dist</li>
    <li>pgadmin4.dist</li>
    <li>python311Packages.flask-mailman</li>
    <li>python311Packages.flask-mailman.dist</li>
    <li>python311Packages.flask-security-too</li>
    <li>python311Packages.flask-security-too.dist</li>
    <li>python311Packages.pony</li>
    <li>python311Packages.pony.dist</li>
    <li>python311Packages.ponywhoosh</li>
    <li>python311Packages.ponywhoosh.dist</li>
    <li>python312Packages.flask-mailman</li>
    <li>python312Packages.flask-mailman.dist</li>
    <li>python312Packages.flask-security-too</li>
    <li>python312Packages.flask-security-too.dist</li>
    <li>python312Packages.pony</li>
    <li>python312Packages.pony.dist</li>
    <li>python312Packages.ponywhoosh</li>
    <li>python312Packages.ponywhoosh.dist</li>
  </ul>
</details>

**Comment by gador** [comment]
Actually, it seems `agda-pkg` is being unmaintained upstream (https://github.com/agda/agda-pkg/issues/49). Since no pkgs inside `nixpkgs` seem to depend on it and this does not appear to be a simple fix, I'll vote to remove this package.  @alexarice you are listed as maintainer of this package: What do you recommend?

**Comment by gador** [comment]
`tribler` fails due to a dependency failure of `libtorrent-rasterbar-1.2.11` (also a `distutils` python3.12 issue). I'd rather keep the scope of the PR down (as it fixes packages with some dependencies, alongside `pgadmin4`). Maybe some of the maintainers of `tribler` can have a look into it? (libtorrent-rasterbar is unmaintained)

**Review by alexarice — APPROVED** [review]

**Comment by alexarice** [comment]
I think its safe to remove agda-pkg if it isn't maintained upstream. I believe it isn't widely used and is not used in any of the agda infrastructure

**Comment by superherointj** [comment]
I found a fix for `libtorrent-rasterbar-2_0_x`. Add to `pkgs/top-level/all-packages.nix`:

> python = python3Packages.python.withPackages (p: [ p.setuptools ]);

**Comment by gador** [comment]
> python = python3Packages.python.withPackages (p: [ p.setuptools ]);

Cool. But doesn't tribler use a special `libtorrent-rasterbar-1_2_x` version?

**Comment by superherointj** [comment]
> > python = python3Packages.python.withPackages (p: [ p.setuptools ]);
> 
> Cool. But doesn't tribler use a special `libtorrent-rasterbar-1_2_x` version?

Yeah, just mark it broken then.

**Comment by gador** [comment]
> > > python = python3Packages.python.withPackages (p: [ p.setuptools ]);
> > 
> > 
> > Cool. But doesn't tribler use a special `libtorrent-rasterbar-1_2_x` version?
> 
> Yeah. I can mark it broken if you don't want.

Done

**Comment by gador** [comment]
Just FYI: I did try your suggestion with the `1.2` version and it got passed the initial `distutils` error, but later failed to build with 
```
libtorrent-rasterbar> Making all in python
libtorrent-rasterbar> make[2]: Entering directory '/build/source/bindings/python'
libtorrent-rasterbar> /nix/store/m22r7gsvm4w3ib3y2jcsnx0l7mis050q-python3-3.12.4-env/bin/python ./setup.py build
libtorrent-rasterbar> /build/source/bindings/python/src/converters.cpp: In static member function ‘static void to_string_view::construct(PyObject*, boost::python::converter::rvalue_from_python_stage1_data*)’:
libtorrent-rasterbar> /build/source/bindings/python/src/converters.cpp:159:63: error: ‘PyUnicode_AS_DATA’ was not declared in this scope; did you mean ‘PyUnicode_DATA’?
libtorrent-rasterbar>   159 |             data->convertible = new (storage) lt::string_view(PyUnicode_AS_DATA(x)libtorrent-rasterbar>       |                                                               PyUnicode_DATA
libtorrent-rasterbar> /build/source/bindings/python/src/converters.cpp:159:85: error: ‘PyUnicode_GET_DATA_SIZE’ was not declared in this scope
libtorrent-rasterbar>   159 |             data->convertible = new (storage) lt::string_view(PyUnicode_AS_DATA(x), PyUnicode_GET_DATA_SIZE(x));
libtorrent-rasterbar>       |                                                                                     ^~~~~~~~~~~~~~~~~~~~~~~
libtorrent-rasterbar> error: command '/nix/store/kp2j7yn0wzwq5piy494r54dafrh83s6s-gcc-wrapper-13.3.0/bin/g++' failed with exit code 1
```
So, yeah, seems broken either way

**Comment by superherointj** [comment]
Result of `nixpkgs-review pr 325181` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>tribler</li>
  </ul>
</details>
<details>
  <summary>22 packages built:</summary>
  <ul>
    <li>mutmut</li>
    <li>mutmut.dist</li>
    <li>pgadmin4</li>
    <li>pgadmin4-desktopmode</li>
    <li>pgadmin4-desktopmode.dist</li>
    <li>pgadmin4.dist</li>
    <li>python311Packages.flask-mailman</li>
    <li>python311Packages.flask-mailman.dist</li>
    <li>python311Packages.flask-security-too</li>
    <li>python311Packages.flask-security-too.dist</li>
    <li>python311Packages.pony</li>
    <li>python311Packages.pony.dist</li>
    <li>python311Packages.ponywhoosh</li>
    <li>python311Packages.ponywhoosh.dist</li>
    <li>python312Packages.flask-mailman</li>
    <li>python312Packages.flask-mailman.dist</li>
    <li>python312Packages.flask-security-too</li>
    <li>python312Packages.flask-security-too.dist</li>
    <li>python312Packages.pony</li>
    <li>python312Packages.pony.dist</li>
    <li>python312Packages.ponywhoosh</li>
    <li>python312Packages.ponywhoosh.dist</li>
  </ul>
</details>

**Comment by gador** [comment]
Result of `nixpkgs-review pr 325181` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>22 packages built:</summary>
  <ul>
    <li>mutmut</li>
    <li>mutmut.dist</li>
    <li>pgadmin4</li>
    <li>pgadmin4-desktopmode</li>
    <li>pgadmin4-desktopmode.dist</li>
    <li>pgadmin4.dist</li>
    <li>python311Packages.flask-mailman</li>
    <li>python311Packages.flask-mailman.dist</li>
    <li>python311Packages.flask-security-too</li>
    <li>python311Packages.flask-security-too.dist</li>
    <li>python311Packages.pony</li>
    <li>python311Packages.pony.dist</li>
    <li>python311Packages.ponywhoosh</li>
    <li>python311Packages.ponywhoosh.dist</li>
    <li>python312Packages.flask-mailman</li>
    <li>python312Packages.flask-mailman.dist</li>
    <li>python312Packages.flask-security-too</li>
    <li>python312Packages.flask-security-too.dist</li>
    <li>python312Packages.pony</li>
    <li>python312Packages.pony.dist</li>
    <li>python312Packages.ponywhoosh</li>
    <li>python312Packages.ponywhoosh.dist</li>
  </ul>
</details>

**Review by superherointj — APPROVED** [review]

**Comment by superherointj** [comment]
@gador I want to merge this but CI eval isn't helping... Wake me up when this is done.

**Comment by superherointj** [comment]
We should have seen this #325297

**Pull request merged**